File: services/grid-protocols/internal/telemetry/telemetry.go
// Package telemetry wires OpenTelemetry tracing and metrics for gridd.
//
// Configuration is entirely environment driven, matching the platform-wide
// contract:
//
//	OTEL_EXPORTER_OTLP_ENDPOINT  gRPC OTLP collector, e.g. http://localhost:4317.
//	                             Unset means telemetry is disabled (logged loudly
//	                             at boot); an unreachable collector never blocks
//	                             or crashes the service — the gRPC exporters
//	                             connect lazily and retry in the background.
//	OTEL_SERVICE_NAME            resource service.name (default "grid-protocols").
//	OTEL_SERVICE_VERSION         resource service.version (omitted when empty).
//	OTEL_ENVIRONMENT             resource deployment.environment (omitted when empty).
//	OTEL_TENANT_ID               resource tenant.id (default "default").
//	OTEL_SDK_DISABLED=true       escape hatch: forces the no-op SDK even when an
//	                             endpoint is configured.
//
// Metrics are exposed in Prometheus format on the service's own HTTP listener
// (GET /metrics) through the OTel Prometheus exporter; traces go to the
// collector over OTLP gRPC.
package telemetry

import (
	"context"
	"crypto/tls"
	"errors"
	"os"
	"strings"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	promexporter "go.opentelemetry.io/otel/exporters/prometheus"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"google.golang.org/grpc/credentials"
)

// Status reports the telemetry pipeline state for /healthz.
type Status struct {
	Enabled     bool   `json:"enabled"`
	Endpoint    string `json:"endpoint,omitempty"`
	ServiceName string `json:"service_name"`
	Reason      string `json:"reason,omitempty"`
}

// Telemetry owns the SDK providers. The zero-use path is safe: when disabled,
// global providers stay no-op and Shutdown is a no-op.
type Telemetry struct {
	status         Status
	resource       *resource.Resource
	tracerProvider *sdktrace.TracerProvider
	meterProvider  *sdkmetric.MeterProvider
}

// Setup initialises the SDK from the environment. It never fails hard: an
// unparseable endpoint or exporter construction error degrades to disabled
// with a loud log line, because protocol serving must not depend on the
// observability stack. warnf receives exactly one loudly-worded line per
// degradation.
func Setup(defaultServiceName string, warnf func(format string, args ...any)) *Telemetry {
	if warnf == nil {
		warnf = func(string, ...any) {}
	}

	// W3C tracecontext is the propagation contract across the platform. It is
	// installed even when the SDK is disabled so inject/extract code paths are
	// uniform; with no recording spans there is simply nothing to propagate.
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{}, propagation.Baggage{}))

	serviceName := strings.TrimSpace(os.Getenv("OTEL_SERVICE_NAME"))
	if serviceName == "" {
		serviceName = defaultServiceName
	}
	t := &Telemetry{status: Status{ServiceName: serviceName}}

	endpoint := strings.TrimSpace(os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT"))
	switch {
	case strings.EqualFold(os.Getenv("OTEL_SDK_DISABLED"), "true"):
		t.status.Reason = "OTEL_SDK_DISABLED=true"
		warnf("telemetry disabled: OTEL_SDK_DISABLED=true")
		return t
	case endpoint == "":
		t.status.Reason = "OTEL_EXPORTER_OTLP_ENDPOINT not set"
		warnf("telemetry disabled: OTEL_EXPORTER_OTLP_ENDPOINT not set")
		return t
	}

	tenantID := strings.TrimSpace(os.Getenv("OTEL_TENANT_ID"))
	if tenantID == "" {
		tenantID = "default"
	}
	attrs := []attribute.KeyValue{
		semconv.ServiceName(serviceName),
		attribute.String("tenant.id", tenantID),
	}
	if v := strings.TrimSpace(os.Getenv("OTEL_SERVICE_VERSION")); v != "" {
		attrs = append(attrs, semconv.ServiceVersion(v))
	}
	if v := strings.TrimSpace(os.Getenv("OTEL_ENVIRONMENT")); v != "" {
		attrs = append(attrs, semconv.DeploymentEnvironment(v))
	}
	res, err := resource.New(context.Background(), resource.WithAttributes(attrs...))
	if err != nil {
		t.status.Reason = "resource: " + err.Error()
		warnf("telemetry disabled: %s", t.status.Reason)
		return t
	}
	t.resource = res

	target, insecure := grpcTarget(endpoint)
	traceOpts := []otlptracegrpc.Option{otlptracegrpc.WithEndpoint(target)}
	if insecure {
		traceOpts = append(traceOpts, otlptracegrpc.WithInsecure())
	} else {
		traceOpts = append(traceOpts, otlptracegrpc.WithTLSCredentials(
			credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})))
	}
	// The gRPC exporter connects lazily: a collector that is down at boot does
	// not fail construction, and the batcher keeps retrying in the background.
	traceExporter, err := otlptracegrpc.New(context.Background(), traceOpts...)
	if err != nil {
		t.status.Reason = "trace exporter: " + err.Error()
		warnf("telemetry disabled: %s", t.status.Reason)
		return t
	}
	t.tracerProvider = sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(t.tracerProvider)

	// The Prometheus exporter self-registers with the default client_golang
	// registry, so the existing promhttp handler serves it at /metrics.
	// Resource attributes land in target_info by default, which the dashboards
	// do not join on; service.name and tenant.id are flattened onto every
	// series instead, matching the platform's label contract.
	promExporter, err := promexporter.New(
		promexporter.WithResourceAsConstantLabels(
			attribute.NewAllowKeysFilter("service.name", "tenant.id")),
	)
	if err != nil {
		t.status.Reason = "prometheus exporter: " + err.Error()
		warnf("telemetry metrics disabled: %s", t.status.Reason)
	} else {
		t.meterProvider = sdkmetric.NewMeterProvider(
			sdkmetric.WithReader(promExporter),
			sdkmetric.WithResource(res),
		)
		otel.SetMeterProvider(t.meterProvider)
	}

	t.status.Enabled = true
	t.status.Endpoint = endpoint
	return t
}

// Enabled reports whether the SDK is exporting.
func (t *Telemetry) Enabled() bool { return t != nil && t.status.Enabled }

// Status returns the current pipeline state for health reporting.
func (t *Telemetry) Status() Status {
	if t == nil {
		return Status{}
	}
	return t.status
}

// Shutdown flushes and stops the providers. It tolerates a nil receiver and a
// disabled pipeline.
func (t *Telemetry) Shutdown(ctx context.Context) error {
	if t == nil {
		return nil
	}
	var errs []error
	if t.tracerProvider != nil {
		errs = append(errs, t.tracerProvider.Shutdown(ctx))
	}
	if t.meterProvider != nil {
		errs = append(errs, t.meterProvider.Shutdown(ctx))
	}
	return errors.Join(errs...)
}

// grpcTarget splits an OTLP endpoint into a gRPC authority target and whether
// it is plaintext. An endpoint with no scheme is treated as TLS, matching the
// OTLP spec default; http:// selects plaintext.
func grpcTarget(endpoint string) (target string, insecure bool) {
	rest := endpoint
	if i := strings.Index(endpoint, "://"); i >= 0 {
		scheme := strings.ToLower(endpoint[:i])
		rest = endpoint[i+3:]
		return strings.TrimSuffix(rest, "/"), scheme == "http"
	}
	return strings.TrimSuffix(rest, "/"), false
}

File: services/modbus-poller/src/telemetry.rs
//! OpenTelemetry tracing pipeline.
//!
//! The poller keeps its existing JSON `fmt` log output; when telemetry is
//! enabled a `tracing-opentelemetry` layer is added alongside it so every
//! span and `info!`/`warn!`/`error!` event is also exported as OTLP traces.
//!
//! Environment contract (shared with the rest of the platform rollout):
//!
//! - `OTEL_EXPORTER_OTLP_ENDPOINT`: gRPC (tonic) endpoint of an OTLP
//!   collector, e.g. `http://otel-collector:4317`. **Unset or empty →
//!   telemetry is disabled**, the reason is logged loudly at startup, and the
//!   poller runs exactly as before.
//! - `OTEL_SERVICE_NAME`: resource `service.name` (default `modbus-poller`).
//! - `OTEL_SERVICE_VERSION`: resource `service.version` (default: crate
//!   version).
//! - `OTEL_ENVIRONMENT`: resource `deployment.environment.name`; omitted from
//!   the resource when unset.
//! - `OTEL_TENANT_ID`: resource `tenant.id` (default `default`).
//! - `OTEL_SDK_DISABLED=true`: hard escape hatch; disables the SDK no matter
//!   what the endpoint says.
//!
//! The OTLP exporter connects lazily and batches spans on a background
//! worker: an unreachable collector produces export errors in the logs,
//! never a crash, an exit, or a stalled poll cycle.

use std::env;

use opentelemetry::trace::TracerProvider as _;
use opentelemetry::KeyValue;
use opentelemetry_otlp::{SpanExporter, WithExportConfig};
use opentelemetry_sdk::propagation::TraceContextPropagator;
use opentelemetry_sdk::trace::SdkTracerProvider;
use opentelemetry_sdk::Resource;
use tracing_subscriber::registry::Registry;
use tracing_subscriber::Layer;

/// Keeps the tracer provider alive for the process lifetime; `shutdown`
/// flushes buffered spans on a clean exit.
pub struct Telemetry {
    provider: SdkTracerProvider,
}

impl Telemetry {
    /// Flushes and shuts the exporter down. Best effort: a dead collector at
    /// shutdown is logged, not treated as a fatal error.
    pub fn shutdown(self) {
        if let Err(err) = self.provider.shutdown() {
            tracing::warn!(error = %err, "telemetry shutdown did not complete cleanly");
        }
    }
}

/// Outcome of [`init`], logged by `main` once the subscriber exists (anything
/// logged before `tracing_subscriber` init would be dropped silently).
pub enum Status {
    Enabled {
        endpoint: String,
        service_name: String,
    },
    Disabled {
        reason: String,
    },
}

/// Everything the subscriber wiring needs from telemetry setup.
pub struct Init {
    pub layer: Option<Box<dyn Layer<Registry> + Send + Sync>>,
    pub telemetry: Option<Telemetry>,
    pub status: Status,
}

/// Builds the optional OTel layer for the subscriber. The W3C trace-context
/// propagator is installed unconditionally so outbound `traceparent`
/// injection degrades to a no-op (invalid context, nothing injected) when
/// telemetry is disabled.
pub fn init() -> Init {
    opentelemetry::global::set_text_map_propagator(TraceContextPropagator::new());

    let disabled = |reason: &str| Init {
        layer: None,
        telemetry: None,
        status: Status::Disabled {
            reason: reason.to_string(),
        },
    };

    if env::var("OTEL_SDK_DISABLED")
        .map(|v| v == "true")
        .unwrap_or(false)
    {
        return disabled("OTEL_SDK_DISABLED=true");
    }

    let endpoint = match env::var("OTEL_EXPORTER_OTLP_ENDPOINT") {
        Ok(endpoint) if !endpoint.trim().is_empty() => endpoint,
        _ => return disabled("OTEL_EXPORTER_OTLP_ENDPOINT is not set"),
    };

    let exporter = match SpanExporter::builder()
        .with_tonic()
        .with_endpoint(endpoint.clone())
        .build()
    {
        Ok(exporter) => exporter,
        Err(err) => return disabled(&format!("OTLP exporter could not be built: {err}")),
    };

    let service_name =
        env::var("OTEL_SERVICE_NAME").unwrap_or_else(|_| "modbus-poller".to_string());
    let service_version =
        env::var("OTEL_SERVICE_VERSION").unwrap_or_else(|_| env!("CARGO_PKG_VERSION").to_string());
    let tenant_id = env::var("OTEL_TENANT_ID").unwrap_or_else(|_| "default".to_string());

    let mut attributes = vec![
        KeyValue::new("service.version", service_version),
        KeyValue::new("tenant.id", tenant_id),
    ];
    if let Ok(environment) = env::var("OTEL_ENVIRONMENT") {
        if !environment.trim().is_empty() {
            attributes.push(KeyValue::new("deployment.environment.name", environment));
        }
    }
    let resource = Resource::builder()
        .with_service_name(service_name.clone())
        .with_attributes(attributes)
        .build();

    let provider = SdkTracerProvider::builder()
        .with_batch_exporter(exporter)
        .with_resource(resource)
        .build();
    let tracer = provider.tracer(env!("CARGO_PKG_NAME"));
    let layer = tracing_opentelemetry::layer().with_tracer(tracer);

    Init {
        layer: Some(Box::new(layer)),
        telemetry: Some(Telemetry { provider }),
        status: Status::Enabled {
            endpoint,
            service_name,
        },
    }
}
